### PR TITLE
Feature/runtime

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Runtime with some functions implemented.
Based on Filecoin spec and Lotus implementation.
Since specs are not exhausting further enhancement is required.

### Usage Examples or Tests *[optional]*
Added mocks for all dependent classes using pimpl principle.

### CODE REVIEW

File: core/vm/actor/CMakeLists.txt [hunk 1]
```diff
@@ -8,7 +8,7 @@ add_library(actor
     init_actor.cpp
     cron_actor.cpp
     invoker.cpp
-    )
+    ../message/message.hpp)
```
Comment (masterjedy) [inline core/vm/actor/CMakeLists.txt]: fix pls

File: core/vm/actor/actor.hpp [hunk 2]
```diff
@@ -30,7 +30,7 @@ namespace fc::vm::actor {
    * associated with methods in Filecoin VM Actors.
    */
   struct MethodNumber {
-    int method_number;
+    unsigned long long method_number;
```
Comment (masterjedy) [inline core/vm/actor/actor.hpp]: you mean uint64_t?

File: core/vm/runtime/impl/runtime_impl.cpp [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+                                                        gas_available_,
+                                                        method_number,
+                                                        params});
+  OUTCOME_TRY(serialized_message, fc::codec::cbor::encode(*message));
+  OUTCOME_TRY(
+      chargeGas(kOnChainMessageBaseGasCost
+                + serialized_message.size() * kOnChainMessagePerByteGasCharge));
+
+  BigInt gas_cost = gas_price_ * gas_available_;
+  BigInt total_cost = gas_cost + value;
```
Comment (turuslan) [inline core/vm/runtime/impl/runtime_impl.cpp]: Line 128 `OUTCOME_TRY(transfer(from_actor, to_actor, value));` already deducted `value`

Comment (Alexey-N-Chernyshov) [reply core/vm/runtime/impl/runtime_impl.cpp]: fixed

File: core/vm/actor/cron_actor.cpp [hunk 4]
```diff
@@ -6,24 +6,30 @@
...
-      OUTCOME_TRY(vmctx.send(entry.to_addr, entry.method_num, BigInt(0), {}));
+      OUTCOME_TRY(
+          runtime->send(entry.to_addr, entry.method_num, {}, BigInt(0)));
     }
     return outcome::success();
   }
 
   ActorExports CronActor::exports = {
-      {2, ActorMethod(CronActor::epochTick)},
+      {MethodNumber{2}, ActorMethod(CronActor::epochTick)},
```
Comment (masterjedy) [inline core/vm/actor/cron_actor.cpp]: no magic numbers should present. pls consider making a enum and accessing method number by name.

File: core/vm/runtime/runtime.hpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    virtual std::shared_ptr<Indices> getCurrentIndices() const = 0;
+
+    /** Look up the code ID of a given actor address. */
+    virtual outcome::result<CodeId> getActorCodeID(
+        const Address &address) const = 0;
+
+    /**
+     * Send allows the current execution context to invoke methods on other
+     * actors in the system
+     * @param to_address
```
Comment (masterjedy) [inline core/vm/runtime/runtime.hpp]: fulfill description

File: core/vm/runtime/runtime.hpp [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+     * @param address - address of actor to delete
+     */
+    virtual outcome::result<void> deleteActor(const Address &address) = 0;
+
+    /**
+     * @brief Returns IPFS datastore
+     */
+    virtual std::shared_ptr<IpfsDatastore> getIpfsDatastore() = 0;
+
+    virtual std::shared_ptr<UnsignedMessage> getMessage() = 0;
```
Comment (masterjedy) [inline core/vm/runtime/runtime.hpp]: add description

File: core/vm/runtime/runtime_types.hpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace fc::vm::runtime {
+
+  using actor::MethodNumber;
+  using actor::MethodParams;
+  using common::Buffer;
+  using exit_code::ExitCode;
+  using primitives::BigInt;
+  using primitives::address::Address;
+
```
Comment (masterjedy) [inline core/vm/runtime/runtime_types.hpp]: here and below add descriptions to types

Comment (Alexey-N-Chernyshov) [reply core/vm/runtime/runtime_types.hpp]: done

File: core/vm/runtime/runtime.hpp (hunk 6 above)
Comment (Alexey-N-Chernyshov) [reply core/vm/runtime/runtime.hpp]: done

File: core/vm/runtime/runtime_types.hpp (hunk 7 above)
Comment (Alexey-N-Chernyshov) [reply core/vm/runtime/runtime_types.hpp]: done

File: core/vm/runtime/runtime.hpp (hunk 5 above)
Comment (Alexey-N-Chernyshov) [reply core/vm/runtime/runtime.hpp]: ok

File: core/vm/actor/cron_actor.cpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply core/vm/actor/cron_actor.cpp]: ok

**Pull request merged**